src/native_mux_impl/stdio.rs

```diff
@@ -28,25 +29,50 @@ fn get_null_fd() -> Result<RawFd, Error> {
 }
 
 pub(crate) enum Fd {
-    PipeReadEnd(PipeRead),
-    PipeWriteEnd(PipeWrite),
-
+    Owned(OwnedFd),
     Borrowed(RawFd),
     Null,
 }
 
+fn cvt(ret: c_int) -> io::Result<c_int> {
+    if ret == -1 {
+        Err(io::Error::last_os_error())
+    } else {
+        Ok(ret)
+    }
+}
+
+fn set_blocking(fd: RawFd) -> io::Result<()> {
+    let flags = cvt(unsafe { fcntl(fd, F_GETFL) })?;
+    cvt(unsafe { fcntl(fd, F_SETFL, flags & (!O_NONBLOCK)) })?;
+
+    Ok(())
+}
+
 impl Fd {
     pub(crate) fn as_raw_fd_or_null_fd(&self) -> Result<RawFd, Error> {
         use Fd::*;
 
-        match self {
-            PipeReadEnd(fd) => Ok(AsRawFd::as_raw_fd(fd)),
-            PipeWriteEnd(fd) => Ok(AsRawFd::as_raw_fd(fd)),
+        let fd = match self {
+            Owned(owned_fd) => Some(owned_fd.as_raw_fd()),
+            Borrowed(rawfd) => Some(*rawfd),
+            Null => None,
+        };
 
-            Borrowed(rawfd) => Ok(*rawfd),
-            Null => get_null_fd(),
+        if let Some(fd) = fd {
+            set_blocking(fd).map_err(Error::ChildIo)?;
```

Comment:
I think this is only okay to do for file descriptors we know we're going to give away. It'd be bad to do this for a file descriptor that we also continued using under `tokio`. Since this only takes `&self`, I don't actually know that we _know_ the return value will _only_ be given away after this function is called. Another alternative would be to `dup` the FD, set blocking on the dup'd FD, and then give that away.

Reply:
`Command::{stdin, stdout, stderr}` all takes `Stdio` by value.
And `Stdio` can be either `null`, `pipe` or constructed from a raw fd, which is considered taken by value, so it should be fine.

> Another alternative would be to dup the FD, set blocking on the dup'd FD, and then give that away.

`O_NONBLOCK` is shared by all fd pointing to the same file, so duplicating it has no use.

Reply:
Hmm, but doesn't that imply that we should take `self` here, not `&self`? To ensure we don't accidentally continue using the fd after giving away ownership in this way?

Reply:
Here `Fd` holds the `OwnedFd`, which is the write/read end of the pipe that needs to be held until the `RawFd` is sent to the ssh multiplex server.

Reply:
But isn't this the method that produces the fd that we send to the server? If so, it should be fine for it to take `self`. If it isn't, then that implies that the `fd` is still going to be used going forward, in which case setting !non-blocking here may have unintended consequences.

Reply:
Also, best to avoid marking these discussions as resolved until they actually are :)

Reply:
> But isn't this the method that produces the fd that we send to the server?

It's actually `Stdio::{to_stdin, to_stdout, to_stderr}` that produces the fd (`PipeWrite` and `PipeRead`).
`Fd` holds onto them to keep them alive.

> If it isn't, then that implies that the fd is still going to be used going forward, in which case setting !non-blocking here may have unintended consequences.

I just realized that `Command::spawn` takes a mutable reference of `self`, so the `OwnedFd` of `Stdio` can indeed be reused.
Though they can only be used in `Command::spawn`, so setting !non-blocking should be fine.

Reply:
That feels like a foot-gun we're setting ourselves up for forgetting later and accidentally trying to use the FD in a tokio context at some later point in time, and it causing all sorts of hard-to-track-down problems. I'd definitely prefer for this to take `self`.

Reply:
> I'd definitely prefer for this to take self.

I can't take `self` here, since that would drop the `Fd::OwnedFd`.

Reply:
You should be able to use [its implementation of `IntoRawFd`](https://doc.rust-lang.org/std/os/unix/io/struct.OwnedFd.html#method.into_raw_fd) to avoid that

Reply:
> You should be able to use [its implementation of IntoRawFd](https://doc.rust-lang.org/std/os/unix/io/struct.OwnedFd.html#method.into_raw_fd) to avoid that

But we still need to close that fd after sending them to ssh multiplex server.

Sending them over unix stream in linux is essentially `dup`ing them to another process.
If we don't close them, then they will be leaked.

Reply:
I'm not sure I follow — once we send it to the multiplex server, why do we need to _also_ hold on to it and close it later? Won't the multiplex server close it at the appropriate time?

Reply:
> I'm not sure I follow — once we send it to the multiplex server, why do we need to also hold on to it and close it later? Won't the multiplex server close it at the appropriate time?

We send the fd to multiplex server over unix socket using `SCM_RIGHTS` ancillary data and it essentially acts as `dup` but across process boundary.

Sending a fd over would not automatically close it, that fd is still valid and can be used by the process.

[This blog](https://copyconstruct.medium.com/file-descriptor-transfer-over-unix-domain-sockets-dcbbf5b3b6ec) has more in-depth explanation of it.

Reply:
Hmmm, I believe we were talking past each other.

> why do we need to also hold on to it and close it later? Won't the multiplex server close it at the appropriate time?

After [the session is establied](https://github.com/openssh-rust/openssh/blob/5282f39b71d9750cc294d145c3d82ad57634ae0f/src/native_mux_impl/command.rs#L81), the [fds](https://github.com/openssh-rust/openssh/blob/5282f39b71d9750cc294d145c3d82ad57634ae0f/src/native_mux_impl/command.rs#L68) are still valid and we need to close them if `Stdio::pipe` is used as stdio.

Reply:
Right, so we have something like this:

stdin_write_end, stdin_read_end, stdout_write_end, stdout_read_end (and ditto for stderr), each of which is a different fd. We send the stdin read end and the stdout write end to the multiplex server. We keep the stdin write end (so we can send stuff to stdin) and the stdout read end (so we can read what the multiplex server writes to stdout). But why do we need to keep the stdin read end and the stdout write end? Those _should_ be closed by the multiplex server when the remote process is done with them. We should never need to close those. We _will_ need to close the stdin write end (to signal no more input) and we _may_ close the stdout read end (to signal we don't care about any more of the remote process' output), but those are different FDs. Unless I'm missing something?

Reply:
> But why do we need to keep the stdin read end and the stdout write end?

We don't need it, that's why we need to close it to avoid leaking.
The whole purpose of enum `Fd` is to close these fds to avoid leaking.

> Those should be closed by the multiplex server when the remote process is done with them.

Yes.

> We should never need to close those.

Unfortunately we do, because sending them over unix socket using `SCM_RIGHTS` is essentailly `dup`ing them, so we still need to close the original one.

> We will need to close the stdin write end (to signal no more input) and we may close the stdout read end (to signal we don't care about any more of the remote process' output), but those are different FDs. Unless I'm missing something?

It's just how `SCM_RIGHTS` over unix socket work:

When you send a fd over unix socket using `SCM_RIGHTS` to another process, they **`dup`** the fds in another process's file descriptor table.

It does not close the fd in the process that sent it, you still need to close it manually.

To explain it in rust's term, sending fd over unix socket using `SCM_RIGHTS` takes a `BorrowedFd` not `OwnedFd`.
It takes the fd by reference not ownership so it's still valid after sending it over to the multiplex server.

Reply:
Ah, okay, I see, so because it's dup'd by sending, we need to close it on our end. That seems reasonable, but that also implies to me that we should close it immediately after sending it. And this doesn't enforce that. One way to enforce that invariant and not have the possibility of accidentally doing the wrong thing is:
```rust
fn as_raw_fd_or_null_fd(self) -> Result<(RawFd, Option<OwnedFd>), Error> {
```
that way, the place we call this _has_ to ensure that the Fd will no longer be used anywhere, and is also given the `OwnedFd` so that it'll be dropped once the `RawFd` has been sent. Another alternative would be:
```rust
fn give_fd_to(self, f: impl FnOnce(RawFd) -> Result<(), Error>) -> Result<(), Error> {
```
though that may be a little more annoying if `f` needs to be `async`.

Reply:
> That seems reasonable, but that also implies to me that we should close it immediately after sending it.

That's what [`Command::spawn`](https://github.com/openssh-rust/openssh/blob/5282f39b71d9750cc294d145c3d82ad57634ae0f/src/native_mux_impl/command.rs#L64) is doing now.

It's closed immediately after being sent.

> `fn as_raw_fd_or_null_fd(self) -> Result<(RawFd, Option<OwnedFd>), Error> {`

This api is actually quite similar to `Stdio::{to_stdin, to_stdout, to_stderr}`.

The `Stdio::to_std*` apis returns `Result<(Fd, Option<Child*>), Error>`, then `Command::spawn` calls `as_raw_fd_or_null_fd` to establish an `[RawFd; 3]` array which is then sent to multiplex server using mux-client.

Once done, the `Fd`s would all get dropped.

Reply:
Sure, but _this_ function is prone to misuse, because it changes the filedescriptor to non-blocking without also ensuring it'll never be used in tokio context again. I'd much prefer that we only change it to non-blocking just before we give it away, at a point where we can guarantee that it'll no longer be accessible to our tokio side of things.

Reply:
That makes sense, I would change that right away.

Reply:
I've changed the code so that now `set_blocking` is invoked in `Command::spawn`.
This makes it clear that the fd should not be reused and that it is indeed set to blocking mode before sending to the multiplex server.